### PR TITLE
JIT: Optimize Memmove unrolling for constant src

### PR DESCRIPTION
**Pull request by EgorBo**
`LowerCallMemmove` already does a good job unrolling Memmove. Unfortunately, it leaves some opportunities on the table when the source can be a constant (esp, GPR constant), so we can skip a memory load. The reason is that `LowerCallMemmove` cannot rely on VN for that as it's too late.

```cs
void Test_utf16(Span<char> destination) => "True".CopyTo(destination);
void Test_utf8(Span<byte> destination) => "True"u8.CopyTo(destination);
```
Codegen diff:
```diff
; Method Benchmarks:Test_utf16(System.Span`1[ushort]):this (FullOpts)
       sub      rsp, 40
       mov      rax, bword ptr [rdx]
       mov      ecx, dword ptr [rdx+0x08]
       cmp      ecx, 4
       jl       SHORT G_M39687_IG04
-      mov      rcx, 0x21A00204FFC
-      mov      rdx, qword ptr [rcx]
-      mov      qword ptr [rax], rdx
+      mov      rcx, 0x65007500720054
+      mov      qword ptr [rax], rcx
       add      rsp, 40
       ret      
G_M39687_IG04:
       call     [System.ThrowHelper:ThrowArgumentException_DestinationTooShort()]
       int3     
-; Total bytes of code: 43
+; Total bytes of code: 40


; Method Benchmarks:Test_utf8(System.Span`1[ubyte]):this (FullOpts)
       sub      rsp, 40
       mov      rax, bword ptr [rdx]
       mov      ecx, dword ptr [rdx+0x08]
       cmp      ecx, 4
       jl       SHORT G_M38560_IG04
-      mov      rcx, 0x2C391CC5380      ; static handle
-      mov      edx, dword ptr [rcx]
-      mov      dword ptr [rax], edx
+      mov      dword ptr [rax], 0x65757254
       add      rsp, 40
       ret      
G_M38560_IG04:
       call     [System.ThrowHelper:ThrowArgumentException_DestinationTooShort()]
       int3     
-; Total bytes of code: 41
+; Total bytes of code: 33
```

[Diffs](https://github.com/MihuBot/runtime-utils/issues/693)

### CODE REVIEW

**Comment by EgorBo** [comment]
@MihuBot

**Comment by EgorBo** [comment]
@MihuBot

**Comment by EgorBo** [comment]
@MihuBot

File: src/coreclr/jit/assertionprop.cpp [hunk 1]
```diff
@@ -2590,6 +2590,96 @@ AssertionIndex Compiler::optAssertionIsSubtype(GenTree* tree, GenTree* methodTab
...
+
+    if (len > getUnrollThreshold(Memcpy))
+    {
+        JITDUMP("...length is too big to unroll - bail out.\n");
+        return nullptr;
+    }
+
+    // if GetImmutableDataFromAddress returns true, it means that the src is a read-only constant.
+    // Thus, dst and src do not overlap (if they do - it's an UB).
+    uint8_t* buffer = getAllocator().allocate<uint8_t>(len);
```
Comment (jakobbotsch) [inline src/coreclr/jit/assertionprop.cpp:2637]:
```suggestion
    uint8_t* buffer = new (this, CMK_AssertionProp) uint8_t[len];
```

File: src/coreclr/jit/assertionprop.cpp [hunk 2]
```diff
@@ -2590,6 +2590,96 @@ AssertionIndex Compiler::optAssertionIsSubtype(GenTree* tree, GenTree* methodTab
...
+
+        // Merge with the previous result.
+        result = result == nullptr ? storeInd : gtNewOperNode(GT_COMMA, TYP_VOID, result, storeInd);
+
+        lenRemaining -= genTypeSize(type);
+    }
+
+    JITDUMP("...optimized into STOREIND(s)!:\n");
+    DISPTREE(result);
+    getAllocator().deallocate(buffer);
```
Comment (jakobbotsch) [inline src/coreclr/jit/assertionprop.cpp:2679]:
```suggestion
```

**Comment by EgorBo** [comment]
PTAL @jakobbotsch @dotnet/jit-contrib 
[Diffs](https://github.com/MihuBot/runtime-utils/issues/693) `(678 improved, 370 regressed)` as expected, there are size "regressions" from the unrollings. Outerloop passed.

**Review by jakobbotsch — APPROVED** [review]

**Comment by xtqqczze** [comment]
@MihuBot -arm64

**Comment by xtqqczze** [comment]
@EgorBo Regressions on arm64?


| Method     | Toolchain               | Mean     | Error   | Ratio |
|----------- |------------------------ |---------:|--------:|------:|
| WriteBools | Main | 604.1 ns | 0.33 ns |  1.00 |
| WriteBools | PR | 477.4 ns | 0.03 ns |  0.79 |

https://github.com/EgorBot/runtime-utils/issues/110#issuecomment-2395178454

| Method     | Toolchain               | Mean     | Error   | Ratio |
|----------- |------------------------ |---------:|--------:|------:|
| WriteBools | Main | 604.1 ns | 0.12 ns |  1.00 |
| WriteBools | PR | 531.0 ns | 0.73 ns |  0.88 |

https://github.com/EgorBot/runtime-utils/issues/116#issuecomment-2399942388

**Comment by EgorBo** [comment]
> @EgorBo Regressions on arm64?

load might be faster than up to 4 instructions to compose a 64bit constant for "FALS" or "TRUE" utf16 strings on arm64, but it's tricky in the real world, it might be under contention. Anyway, we'll see once dotnet/performance run finishes

**Pull request merged**